sherpa-onnx/csrc/online-conformer-transducer-model.h

```diff
@@ -0,0 +1,108 @@
+#endif
+
+#include "onnxruntime_cxx_api.h"  // NOLINT
+#include "sherpa-onnx/csrc/online-transducer-model-config.h"
+#include "sherpa-onnx/csrc/online-transducer-model.h"
+
+namespace sherpa_onnx {
+
+class OnlineConformerTransducerModel : public OnlineTransducerModel {
+ public:
+  explicit OnlineConformerTransducerModel(
+      const OnlineTransducerModelConfig &config);
+
+#if __ANDROID_API__ >= 9
+  OnlineConformerTransducerModel(AAssetManager *mgr,
+                                 const OnlineTransducerModelConfig &config);
+#endif
+
+  std::vector<Ort::Value> StackStates(
+      const std::vector<std::vector<Ort::Value>> &states) const override;
+
+  std::vector<std::vector<Ort::Value>> UnStackStates(
+      const std::vector<Ort::Value> &states) const override;
+
+  std::vector<Ort::Value> GetEncoderInitStates() override;
+
+  std::pair<Ort::Value, std::vector<Ort::Value>> RunEncoder(
+      Ort::Value features, std::vector<Ort::Value> states,
+      Ort::Value processed_frames) override;
+
+  Ort::Value RunDecoder(Ort::Value decoder_input) override;
+
+  Ort::Value RunJoiner(Ort::Value encoder_out, Ort::Value decoder_out) override;
+
+  int32_t ContextSize() const override { return context_size_; }
+
+  int32_t ChunkSize() const override { return T_; }
+
+  int32_t ChunkShift() const override { return decode_chunk_len_; }
+
+  int32_t VocabSize() const override { return vocab_size_; }
+  OrtAllocator *Allocator() override { return allocator_; }
+
+ private:
+  void InitEncoder(void *model_data, size_t model_data_length);
+  void InitDecoder(void *model_data, size_t model_data_length);
+  void InitJoiner(void *model_data, size_t model_data_length);
+
+ private:
+  Ort::Env env_;
+  Ort::SessionOptions sess_opts_;
+  Ort::AllocatorWithDefaultOptions allocator_;
+
+  std::unique_ptr<Ort::Session> encoder_sess_;
+  std::unique_ptr<Ort::Session> decoder_sess_;
+  std::unique_ptr<Ort::Session> joiner_sess_;
+
+  std::vector<std::string> encoder_input_names_;
+  std::vector<const char *> encoder_input_names_ptr_;
+
+  std::vector<std::string> encoder_output_names_;
+  std::vector<const char *> encoder_output_names_ptr_;
+
+  std::vector<std::string> decoder_input_names_;
+  std::vector<const char *> decoder_input_names_ptr_;
+
+  std::vector<std::string> decoder_output_names_;
+  std::vector<const char *> decoder_output_names_ptr_;
+
+  std::vector<std::string> joiner_input_names_;
+  std::vector<const char *> joiner_input_names_ptr_;
+
+  std::vector<std::string> joiner_output_names_;
+  std::vector<const char *> joiner_output_names_ptr_;
+
+  OnlineTransducerModelConfig config_;
+
+  int32_t num_encoder_layers_ = 0;
+  int32_t T_ = 0;
+  int32_t decode_chunk_len_ = 0;
+  int32_t cnn_module_kernel_ = 0;
+  int32_t context_size_ = 0;
+  int32_t left_context_ = 0;
+  // TODO: to get from model
+  int32_t right_context_ = 4;
+  int32_t encoder_dim_ = 0;
+  int32_t pad_length_ = 0;
+  int32_t vocab_size_ = 0;
```

Comment:
Except for `right_context` (missing in the model metadata), all the other parameters are retrieved from the model. I recommended earlier to add `right_context` to the ONNX export script. 

```cpp
  SHERPA_ONNX_READ_META_DATA(left_context_, "left_context");
  SHERPA_ONNX_READ_META_DATA(encoder_dim_, "encoder_dim");
  SHERPA_ONNX_READ_META_DATA(pad_length_, "pad_length");
  SHERPA_ONNX_READ_META_DATA(cnn_module_kernel_, "cnn_module_kernel");
......
  SHERPA_ONNX_READ_META_DATA(vocab_size_, "vocab_size");
```

Reply:
@jingzhaoo 

I just updated the encoder model to include `right_context` in its metadata.

By the way, I don't think you need right_context in sherpa-onnx.

Is this PR ready to merge?

Reply:
@csukuangfj thanks for adding `right_context` to the model metadata. Look like both `left_context` and `right_contex` are not required in `sherpa-onnx`. 

I just committed all I have into the PR. Things are not working yet. I think stacking and unstacking states are not working as expected. I ran into segmentation fault:

```
===> StackStates (start)                                                                                                                    
===> StackStates (end)                                                                                                                      
state: 18,64,1,512,                                                                                                                         
state: 18,30,1,512,                                                                                                                         
===> model_->RunEncoder (begin)                                                                                                             
===> model_->RunEncoder (end)                                                                                                               
===> model_->Decoder (begin)                                                                                                                
===> model_->Decoder (end)                                                                                                                  
===> model_->UnStackStates (begin)                                                                                                          
===> model_->UnStackStates (end)                                                                                                            
===> set (start: 1, 1, 64)                                                                                                                  
===> set (end)                                                                                                                              
Segmentation fault (core dumped) 
```

I jhate to hold off any progress of this project. I suggest that we can merge the PR (or through other means) so that you can debug the parts that I get wrong. Since this is a new feature, I don't see it breaking any existing features. Please let me know what you suggest. Thanks.

Reply:
When the code runs to 

```cpp
    for (int32_t i = 0; i != n; ++i) {
      ss[i]->SetResult(results[i]);
      ss[i]->SetStates(std::move(next_states[i]));
    }
```

`next_states.size()` is 64, rather than 1. Thus, the output of `next_states = model_->UnStackStates(pair.second)` is incorrect due to some reason. I am still thinking why that is the case. Not much clue yet.

Reply:
> I jhate to hold off any progress of this project. I suggest that we can merge the PR (or through other means) so that you can debug the parts that I get wrong.

No worry. I am helping debug it. I will try to find out what goes wrong.

Reply:
@csukuangfj Thanks a lot! I looked at `pair = model_->RunEncoder(...)`. `pair.second` has two `Ort::Value`s with shapes of `(18, 64, 1, 512)` and `(18, 30, 1, 512)`, which should be correct based on my understanding of the model init process. 

After the following two `Unbind`s:

```cpp
  std::vector<Ort::Value> attn_vec = Unbind(allocator_, &states[0], 2);
  std::vector<Ort::Value> conv_vec = Unbind(allocator_, &states[1], 2);
```

`attn_vec[0]` has shape `(18, 64, 1, 512)` while `conv_vec[0]` has shape `(18, 30, 1, 512)`. This seems incorrect. `Unbind` has a different behavior from `torch::unbind`. I suspect we may need to write a different `Unbind` function here.

Reply:
@jingzhaoo 

I just realized there is not need to implement `Stack()`.
We can change the init states to a 4-D tensor and use Cat() to stack them.

---

Please see
https://github.com/jingzhaoou/sherpa-onnx/pull/1


---

Also, please remove the debug statements so that we can merge it.

Reply:
@csukuangfj Indeed! After merging your PR to my branch, things are working correctly. You are so awesome! I am cleaning my debug statements now. 

The stack code can stay just in case we need it later on. Let me know if you think otherwise.

Reply:
> The stack code can stay just in case we need it later on.

Yes, please keep the stack code.